ophyd/signal.py

```diff
@@ -810,7 +921,116 @@ def limits(self):
         return (self._metadata['lower_ctrl_limit'],
                 self._metadata['upper_ctrl_limit'])
 
-    def get(self, *, as_string=None, connection_timeout=1.0, form='time',
+    def _get_with_retries(self, pv, timeout, connection_timeout, retries,
+                          as_string, form):
+        """
+        Utility method implementing a retry loop for get and get_setpoint
+
+        Returns info from pv.read_with_metadata(...) or raises TimeoutError
+        """
+        # Fall back to instance default if no value is given.
+        if timeout is DEFAULT_TIMEOUT:
+            timeout = self.timeout
+        if connection_timeout is DEFAULT_CONNECTION_TIMEOUT:
+            connection_timeout = self.connection_timeout
+        if retries is DEFAULT_READ_RETRIES:
+            retries = self.read_retries
+        retries = int(retries)  # normalize/validate input
+        floor = self.__read_attempt_timeout_floor  # alias for brevity
+        attempts = 1 + retries
+        start_time = time.monotonic()
+        if timeout is None:
+            # The user has told us to wait forever. How shall we divide
+            # "forever" up into retries? Let us choose 100X the floor timeout
+            # per read.  This is a pretty weird edge case not worth worry about
+            # much: users really shouldn't ask us to wait forever.
+            read_timeout = floor * 100
+            deadline = None
+        else:
+            # Split up our total time budget (timeout) into the max number of
+            # attempts. The max(...) ensures that each attempt is given a
+            # reasonable chance to succeed by never making each one smaller
+            # than `floor`.  The min(...) ensures that, in the edge case where
+            # the user specifically asks for a timeout less than the floor, we
+            # respect that.
+            timeout = float(timeout)
+            read_timeout = min(timeout, max(timeout / attempts, floor))
+            deadline = start_time + timeout
+        if connection_timeout is not None:
+            connection_timeout = float(connection_timeout)
+
+        # Retry connecting and reading until either our entire time budget
+        # (timeout) is exhausted or our number of read_retries is reached,
+        # whichever happens first.
+        err = None
+        for attempt in range(attempts):
+            try:
+                try:
+                    self.wait_for_connection(timeout=connection_timeout)
+                except TimeoutError:
+                    raise ConnectionTimeoutError(
+                        f"Failed to connect to {self._read_pvname} "
```

Comment:
We need to ask the `pv` object it's name rather than using `self._read_pvname` as this code path is used from both the readback and setpoint PVs.

Reply:
Good catch.